pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -215,74 +216,104 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
 	pdbStateMap *corev1.ConfigMap,
 	failureDomainType string,
 	allFailureDomains,
-	drainingFailureDomains []string,
-) error {
+	osdDownFailureDomains []string,
+	activeNodeDrains bool,
+) (reconcile.Result, error) {
+	var osdDown bool
+	var drainingFailureDomain string
+	if len(osdDownFailureDomains) > 0 {
+		osdDown = true
+		drainingFailureDomain = osdDownFailureDomains[0]
+	}
 
-	pgHealthMsg, clean, err := cephclient.IsClusterClean(r.context.ClusterdContext, clusterInfo)
+	pgHealthMsg, pgClean, err := cephclient.IsClusterClean(r.context.ClusterdContext, clusterInfo)
 	if err != nil {
 		// If the error contains that message, this means the cluster is not up and running
 		// No monitors are present and thus no ceph configuration has been created
 		if strings.Contains(err.Error(), opcontroller.UninitializedCephConfigError) {
-			logger.Debugf("Ceph %q cluster not ready, cannot check Ceph status yet.", request.Namespace)
-			return nil
+			logger.Infof("Ceph %q cluster not ready, cannot check Ceph status yet.", request.Namespace)
+			return opcontroller.WaitForRequeueIfOperatorNotInitialized, nil
 		}
-		return errors.Wrapf(err, "failed to check cluster health")
-	}
-
-	if pdbStateMap.Data == nil {
-		pdbStateMap.Data = make(map[string]string)
-	}
-	_, ok := pdbStateMap.Data[drainingFailureDomainKey]
-	if !ok {
-		pdbStateMap.Data[drainingFailureDomainKey] = ""
+		return reconcile.Result{}, errors.Wrapf(err, "failed to check cluster health")
 	}
 
-	var drainingFailureDomain string
-	activeDrains := len(drainingFailureDomains) > 0
-
-	// create blocking PDBs for all the OSDs that are not being drained right now
-	if activeDrains {
-		drainingFailureDomain = drainingFailureDomains[0]
-		logger.Infof("all %q failure domains: %v. currently draining: %q. pg health: %q", failureDomainType,
-			allFailureDomains, drainingFailureDomain, pgHealthMsg)
-		err := r.handleActiveDrains(allFailureDomains, drainingFailureDomain, failureDomainType, clusterInfo.Namespace, clean)
+	switch {
+	// osd is down but pgs are active+clean
+	case osdDown && pgClean:
+		lastDrainTimeStamp, err := getLastDrainTimeStamp(pdbStateMap, drainingFailureDomainDurationKey)
 		if err != nil {
-			return errors.Wrap(err, "failed to handle active drains")
+			return reconcile.Result{}, errors.Wrapf(err, "failed to get last drain timestamp from the configmap %q", pdbStateMap.Name)
+		}
+		timeSinceOSDDown := time.Since(lastDrainTimeStamp)
+		if timeSinceOSDDown > 30*time.Second {
+			logger.Infof("osd is down in failure domain %q is down for the last %.2f minutes, but pgs are active+clean", drainingFailureDomain, timeSinceOSDDown.Minutes())
+			pdbStateMap.Data[drainingFailureDomainKey] = ""
+			delete(pdbStateMap.Data, drainingFailureDomainDurationKey)
+			delete(pdbStateMap.Data, pgHealthCheckDurationKey)
+		} else {
+			logger.Infof("osd is down in the failure domain %q, but pgs are active+clean. Requeuing...", drainingFailureDomain)
```

Comment:
```suggestion
			logger.Infof("osd is down in the failure domain %q, but pgs are active+clean. Requeuing in case pg status is not updated yet...", drainingFailureDomain)
```